doc/dumper.1
.\" Automatically generated by Pandoc 3.2
.\"
.TH "dumper" "1" "" "" "General Commands Manual"
.SH NAME
dumper \- a multi\-radix file dump program
.SH SYNOPSIS
\f[CR]dumper [\-c \-n \-r \-v] [\-\-help] [[\-d \-h \-o] [file ...]]\f[R]
.SH DESCRIPTION
For each operand that refers to a regular file, or a symbolic or hard
link to a regular file, dumper displays a numeric and printable
character representation of the bytes in the file, sequentially.
This format is often referred to as a radix dump.
If no operands are given, dumper accepts input from stdin.
When the special name \[lq]\-\[rq] is given as an argument, dumper reads
from stdin until end of file.
All of the files named on the command line are processed in the order
specified.
The last \f[B]\-d\f[R], \f[B]\-h\f[R], or \f[B]\-o\f[R] that precedes a
file name controls the radix for the dump of that file.
.PP
The following options are available:
.TP
\f[CR]\-c\f[R]
columns alternate with highlights using VT/ANSI escape codes
.TP
\f[CR]\-d\f[R]
dump using decimal numbers
.TP
\f[CR]\-h\f[R]
dump using hexadecimal numbers; this is the default
.TP
\f[CR]\-\-help\f[R]
a synonym for \-v
.TP
\f[CR]\-n\f[R]
emit the name of a file on a separate line immediately before the start
of its content dump
.TP
\f[CR]\-o\f[R]
dump using octal numbers
.TP
\f[CR]\-r\f[R]
reset the offset counter for each file when dumping multiple files; the
default is to keep a running offset as if all files were concatenated
.TP
\f[CR]\-v\f[R]
display version information and usage synopsis
.SH ENVIRONMENT
The C Library may be locale\-aware, in which case the standard locale
environment variables may affect the processing of bytes for ASCII
output.
The behavior with collating sequences other than US ASCII is not
specified.
.SH EXIT STATUS
The dumper utility exits 0 on success, and >0 if an error occurs.
.SH EXAMPLES
Dump the contents of the file /bin/ls in hexadecimal mode, without
column highlighting:
.IP
.EX
$ dumper /bin/ls
.EE
.PP
Dump the contents of stdin, /bin/ls, and /etc/motd with different
radices:
.IP
.EX
$ dumper \-d \- \-o /bin/ls \-h /etc/motd
.EE
.PP
Dump the contents of stdin, /bin/ls, and /etc/motd with different
radices, resetting the offset for each file, emitting the name of each
file, and highlighting alternating columns in the radix dump:
.IP
.EX
$ dumper \-d \- \-o /bin/ls \-h /etc/motd \-c \-r \-n
.EE
.SH SEE ALSO
hexdump(1), od(1), xxd(1)
.SH HISTORY
A dumper command appeared on select SunOS 4.1, NeXTStep 2.0, and MS\-DOS
3.x\-6.x systems at Rose\-Hulman Institute of Technology in 1991.
.SH BUGS
The behavior on very large files is not known.
.SH LICENSE
Copyright 1991\-2024 Jillian Alana Bolton
.PP
The BSD 2\-Clause License
.PP
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:
.IP "1." 3
Redistributions of source code must retain the above copyright notice,
this list of conditions and the following disclaimer.
.IP "2." 3
Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution.
.PP
THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS \[lq]AS IS\[rq]
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED.
IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.


CHANGES.md
Changelog for a multi-radix file dump program
=============================================

License
-------

Copyright 1991-2024 Jillian Alana Bolton

The BSD 2-Clause License

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above
    copyright notice, this list of conditions and the following
    disclaimer in the documentation and/or other materials provided
    with the distribution.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS
BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes Summary for 5 June 2024 by Jillian Alana Bolton
-------------------------------------------------------

- Added support for multiple files, each with different radices
- Added support for stdin when using UNIX I/O
- Added -c switch for column highlights
- Added -r switch for resetting the offset between files
- Added --help switch as a synonym for -v
- Updated usage message
- Cleaned up default UNIX makefile and removed "install" target
- Print filenames as given (no backslash conversion and strlwr)
- General cleanup of various cruft, comments, and documentation
- Moved VMS and Turbo C specific files to the attic
- Changed function definitions to ANSI style
- Patterned inbuilt strerror() after POSIX strerror_r(3)
- Moved the old README to attic
- created balun.h, dumper.h, and moved stuff necessary to compile
  as a library into there (left as an exercise for the reader)
- Created new README from the ashes of the old
- Added Visual Studio project files

Changes Summary for unknown date in 1995 by Mark P. Fister
----------------------------------------------------------

- Added alternting VT / ANSI highlights in the hex, dec, oct columns

Changes Summary for 25 January 1994 by Jillian Alana Bolton
------------------------------------------------------------

-  Converted error messages to standard UNIX-style messages
-  Added code to convert upper case paths in MS-DOS to lower case
-  Altered error messages to include strerror output
-  Added open,read,close for better VMS compatibility
-  Added makefile for Turbo C with the following targets:

`dumper`
: just the executable file.

`clean`
: remove the executable, object files, and backups.

`install`
: builds dumper and copies it to BINDIR

Changes Summary for 28 February 1994 by Jillian Alana Bolton
------------------------------------------------------------

- Changed command line format and usage message:  switches now go before
  the filename
  Thanks to Ade Barkah (mbarkah@barkah.golden.co.us) for the suggestion
- stdin is still not supported.     
- Changed UNIX Makefile for compatibility with make tools requiring tabs
- Changed UNIX Makefile so default target is dumper rather than clean
  Thanks again, Ade. :)
- Fixed bugs in the preprocessor spaghetti
  -DNO_STRERROR, -DNO_ERRNO, -DNO_SYS_ERRLIST now work as intended

Changes Summary for 1 March 1994 by Jillian Alana Bolton
--------------------------------------------------------

- Now using perror for errors, and conditional identifier -DNO_PERROR
- Added -DBSD identifier to include either <string.h> or <strings.h>
  (Thanks again, Ade, for the suggestions)
- Fixed a bug in the local strerror() code that used errno explicitly
- Fixed an out of range index into sys_errlist in strerror

Changes Summary for 2 March 1994 by Jillian Alana Bolton
--------------------------------------------------------

- Fixed VMS port so that it works as described
- UNIX I/O routines are selected at build time with -DUSE_LOW_LEVEL_IO
- The default build uses stream I/O from stdio.h
- Redefined EXIT_FAILURE for VMS versions so that you don't get a
  system message anymore since dumper prints its own error messages
- stdin is supported when USE_LOW_LEVEL_IO is not defined.

Changes Summary for 17 February 1995 by Jillian Alana Bolton
------------------------------------------------------------
- Changed function definitions from ANSI style to K & R style
  for better portability.
- Updated README file with more systems.


dumper.c
/* dumper.c - a multi-radix file dump program
 *
 * Copyright 1991-2024 Jillian Alana Bolton
 * Copyright 1995 Mark P. Fister
 *
 * The BSD 2-Clause License
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *   1. Redistributions of source code must retain the above copyright
 *      notice, this list of conditions and the following disclaimer.
 *
 *   2. Redistributions in binary form must reproduce the above
 *      copyright notice, this list of conditions and the following
 *      disclaimer in the documentation and/or other materials provided
 *      with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
 * OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
 * IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "balun.h"
#include "dumper.h"

#define COMPILATION     "Compiled"
#define DESCRIPTION     "dumper - a file dump program."
#define HEX_BLANK       "+-"
#define NON_PRINT_CHAR  '.'
#define OTHER_BLANK     "+--"
#define RESTRICTIONS    "\nSee documentation and source for distribution restrictions."
#define USAGE_MSG       "%s:  usage:\n\t%s [-c -n -r -v] [--help] [[-d -h -o] [file ...]]\n"
#define VERSION         "\nJillian Alana Bolton\nReleased Mar 02 1994 07:00:00"

/* constants for command line switches. */

#define COL             "-c"
#define DASH            "-"
#define DEC             "-d"
#define DEFAULT_MODE    hexadecimal
#define HELP            "--help"
#define HEX             "-h"
#define NAM             "-n"
#define OCT             "-o"
#define RST             "-r"
#define VER             "-v"

#define BUFLEN           16 * sizeof(char)

typedef struct {
    int            handle;
    FILE          *stream;
    modetype       mode;
    const char    *name;
} filetype;

static const char *stdinname = "stdin";

void dumper(char *s, size_t c, size_t *offset, modetype mode, int h)
{
    int loop;
    int column = 2;
    unsigned char *p = (void *)s;

    switch (mode) {
	case hexadecimal:
	    printf("%08lx: ", *offset);
	    break;
	case octal:
	    printf("%010lo: ", *offset);
	    break;
	case decimal:
	    printf("%010lu: ", *offset);
	    break;
    }

    /* offset is the index into the file of the first char in the buffer.
     * Ideally, it's a multiple of 16, but on VMS, read() and fread() may
     * stop early, i.e. at the end of each record. In such cases, ASCII
     * EOL characters are not present.
     */

    *offset += c;

    for (loop = 0; loop <= (c - 1); loop++) {

            /* separate every group of four numbers in hex mode, 8 otherwise. */

	if ((!(loop % 4) && loop && (mode == hexadecimal)) ||
	    (!(loop % 8) && loop && (mode != hexadecimal)))
	    printf(" ");

	if (h && (++column % 2)) {
	    printf("[7m");
        }

	switch (mode) {
	    case hexadecimal:
		printf("%02x", (unsigned int) p[loop]);
		break;
	    case decimal:
		printf("%03d", (unsigned int) p[loop]);
		break;
	    case octal:
		printf("%03o", (unsigned int) p[loop]);
		break;
	}

	if (h && (column % 2)) {
	    printf("[0m");
        }
            
            /* if it's not printable, print something else. */

	if (!(isalnum(s[loop]) || ispunct(s[loop])
	      || (s[loop] == ' ')))
	    s[loop] = NON_PRINT_CHAR;

    }

    for (; loop < 16; loop++) {

            /* separate every group of four numbers in hex mode, 8 otherwise */

	if ((!(loop % 4) && loop && (mode == hexadecimal)) ||
	    (!(loop % 8) && loop && (mode != hexadecimal)))
	    printf(" ");

            /* print blanks in  leftover positions to fill the number space. */

	printf(((mode == hexadecimal) ? HEX_BLANK : OTHER_BLANK));

    };

    s[c] = 0;

    printf(" %s\n", s);
};

int fileopen(char *fname, filetype *f)
{
    /* VAX C sequential files have associated trade-offs between UNIX I/O and
     * buffered C I/O.  To select UNIX I/O, define USE_LOW_LEVEL_IO
     */
#ifdef USE_LOW_LEVEL_IO
    f->handle = open(fname, O_RDONLY | O_BINARY | O_NDELAY, S_IFREG);
    return (f->handle >= 0);
#else
    f->stream = fopen(fname, "rb");
    return (f->stream != NULL);
#endif
}

void fileclose(filetype *f)
{
#ifdef USE_LOW_LEVEL_IO
    close(f->handle);
#else
    if (f->stream != stdin)
	fclose(f->stream);
#endif
}

int fileread(char *buf, size_t count, filetype *f)
{

#ifdef USE_LOW_LEVEL_IO
    return read(f->handle, buf, (unsigned) count);
#else
    return fread(buf, (size_t) 1, (size_t) count, f->stream);
#endif
}

void emitversion(char *exename)
{
    printf("%s\n", DESCRIPTION);
    printf("%s \n%s %s %s\n%s\n\n", VERSION, COMPILATION, __DATE__, __TIME__, RESTRICTIONS);
    printf(USAGE_MSG, exename, exename);
}

int main(int argc,char *argv[])
{
    int       nread;
    char      errmsg[1024];                 /* an arbitrary limit */
    modetype  mode = DEFAULT_MODE;	    /* What mode are we in? */

    char      inbuf[(BUFLEN + sizeof(char))]; /* and one for the javanisse */
    int       i;
    int       filecount = 0;                 /* how many args were pathnames */
    filetype *files;
    int       columnhighlight = 0;
    int       printnames = 0;
    int       resetoffset = 0;
    size_t    offset = 0;
    
    files = malloc(sizeof(filetype) * argc); /* always one or more due to argv[0] */

    if (!files) {
        perror("out of memory");
        return EXIT_FAILURE;
    }

    for (i = 1; i < argc; i++) {
        if (!strcmp(argv[i], VER) || !strcmp(argv[i], HELP)) {
            emitversion(argv[0]);
            return EXIT_SUCCESS;
        } else if (!strcmp(argv[i], DEC)) {
            mode = decimal;
        } else if (!strcmp(argv[i], OCT)) {
            mode = octal;
        } else if (!strcmp(argv[i], HEX)) {
            mode = hexadecimal;
        } else if (!strcmp(argv[i], RST)) {
            resetoffset = 1;
        } else if (!strcmp(argv[i], NAM)) {
            printnames = 1;
        } else if (!strcmp(argv[i], COL)) {
            columnhighlight = 1;
        } else if (!strcmp(argv[i], DASH)) {
            files[filecount].handle = 0;
            files[filecount].stream = stdin;
            files[filecount].mode = mode;
            files[filecount].name = stdinname;
            filecount++;
        } else {
            if (!fileopen(argv[i], &files[filecount])) {
                sprintf(errmsg, "%s: %s", argv[0], argv[i]);
                perror(errmsg);
                return EXIT_FAILURE;
            }
            files[filecount].mode = mode;
            files[filecount].name = argv[i];
            filecount++;
        }
    }

    if (filecount == 0) {
        files[filecount].handle = 0;
        files[filecount].stream = stdin;
        files[filecount].mode = mode;
        files[filecount].name = stdinname;
        filecount++;
    }

    for (i = 0; i < filecount; i++) {

        if (resetoffset) {
            offset = 0;
        }

        if (printnames) {
            printf("%s:\n",files[i].name);
        }

        while ((nread = fileread(inbuf, (size_t) BUFLEN, &files[i]))) {
            if (nread < 0) {
                    /* some serious error occurred. */
                sprintf(errmsg, "%s: %s", argv[0], files[i].name);
                perror(errmsg);
                fileclose(&files[i]);
                return EXIT_FAILURE;
            }
            inbuf[nread] = 0;
            dumper(inbuf, nread, &offset, files[i].mode, columnhighlight);
        }
        fileclose(&files[i]);
    }

    return EXIT_SUCCESS;
}
